### PR TITLE
SI-7186 Slim down some TypeRefs by 8 bytes.

### PR DESCRIPTION
**Pull request by retronym**
By virtue of being defined in object TypeRef, they
had two outer pointers.

```
$outer  scala.reflect.internal.Types$TypeRef$
$outer  scala.tools.nsc.Global
```

This commit moves them out of that object. They also
now are no longer shrouded in anonymity, which is
pleasant when debugging and profiling.

The shallow size leaderboard now looks like:

```
   Class             % Shallow  % Objects
                          Size
-----------------------------------------
1. ::                      14%        25%
2. char[]                  10% a       5%
3. ClassArgsTypeRef         6%         5%
4. Object[]                 6% b       1%
5. TypeHistory              5%         6%
6. AbstractTypeSymbol       4%         2%
7. AbsractNoArgsTypeRef     3%         2%
8. j.l.Class                3%         2%
9. TermSymbol               2%         1%
10. HashEntry               1% c       1%
```

a) 6% of the char[] by size is the name table.
   Retained source code seems to be sizable => should we discard after parsing?
b) Types.uniques
c) ZipArchive#DirEntry.entries

Confirmation of the slimmer size can be seen in:

  https://github.com/retronym/scala/compare/ticket/7186

In that branch, I introduced infrastructure to query JVM
object size in instrumented tests, and the diff shows:

``` patch
       size                class instance
 ========================================================================
-  64 bytes  internal.Types$TypeRef$$anon$5 List[Int]
-  64 bytes  internal.Types$TypeRef$$anon$6 String
+  56 bytes  nternal.Types$ClassArgsTypeRef List[Int]
+  56 bytes  ernal.Types$ClassNoArgsTypeRef String
```

I chose not to bring that test infrastructure across to master so as
not to burden us with brittle tests built on esoteric facilities.

This does not close the ticket which proposes a different
space optimization which is also explored in that branch,
but no a clear cut improvement.

Review by @gkossakowski (if you're up to it) or @paulp

### CODE REVIEW

**Comment by paulp** [comment]
You reminded me of something I tried out four days ago and forgot about. After noticing 90+% of certain TypeRef fields are null, I [trimmed TypeRefs this way](https://github.com/paulp/scala/compare/topic;typeref-fields). Memory usage was lower, but not as much as I expected. What do you think?

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -2422,20 +2422,28 @@ trait Types
...
   }
 
+  // No longer defined as anonymous classes in `object TypeRef` to avoid an unnecessary outer pointer.
+  private final class AliasArgsTypeRef(pre: Type, sym: Symbol, args: List[Type]) extends ArgsTypeRef(pre, sym, args) with AliasTypeRef
+  private final class AbstractArgsTypeRef(pre: Type, sym: Symbol, args: List[Type]) extends ArgsTypeRef(pre, sym, args) with AbstractTypeRef
+  private final class ClassArgsTypeRef(pre: Type, sym: Symbol, args: List[Type]) extends ArgsTypeRef(pre, sym, args) with ClassTypeRef
+  private final class AliasNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym) with AliasTypeRef
+  private final class AbstractNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym) with AbstractTypeRef
+  private final class ClassNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym) with ClassTypeRef
+
```
Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala:2432]: Yeah, I've meant to do this forever, thanks. At some point I rescued lots of other types from the bonus outer pointer; wonder how I overlooked these. I guess define-and-instantiate-simultaneously foiled my recognition algorithm.

**Comment by paulp** [comment]
LGTM LGTM LGTM

**Comment by retronym** [comment]
I just incorporated your patch into my other branch. The TypeRefs shrink from 54 to 40 bytes. https://github.com/retronym/scala/commit/fc161960

But that would only lead to a small drop in retained memory, only about 9% of retained memory is type refs, so I'd expect about 2-3% total reduction.

Even if that patch only maintains current performance, I'd consider merging it.

**Comment by paulp** [comment]
I guess I'll take "only 3%" !

Can I consider that patch a part of your branch with a future in some PR, so I can delete mine?

**Comment by retronym** [comment]
Well, the rest of that branch has a less certain future, but I won't close SI-7186 without considering your branch.

**Comment by gkossakowski** [comment]
I wish we were a little bit slower at merging PRs so I had a chance to look at it.

**Comment by paulp** [comment]
I will give you a thousand dollars if you can find something to object to in this commit. Anyway, you can still look at it. The wheel spins both ways.

**Comment by retronym** [comment]
@gkossakowski Glad to hear you're keen to review! As Paul noted, this one is pretty uncontroversial. What I'm really after is a methodical and repeatable way to measure the memory footprint. Do you have something like that? I suppose could add an option to trigger a memory dump via JMX at the end of a run (before the per-run-caches are cleared), but I don't want to duplicate effort.

**Comment by adriaanm** [comment]
Reviewing >= objecting. Reviewing ==> sharing knowledge.

**Comment by som-snytt** [comment]
Those long identifiers are pretty unwieldy, especially in a debugger.  At least with anons, I can cut/paste and just change a 1 to a 2, etc.  Since they're private final syms, couldn't they be very short?  AATR?  Short acronym never hurt AARP, which now means An Ally for Real Possibilities.

Unless paulp meant Monopoly dollars, in which case, I retract my objection.

**Comment by paulp** [comment]
I'm not sure I see how merging a pull request limits further knowledge-sharing opportunities. The Door Is Always Open!

**Comment by som-snytt** [comment]
I agree with the sentiment that if I blink I might miss something good. Anything closed is filed away. Github is an amazing tool, but it would be nice if I could browse these discussions as I do my GMail.  Ideally, a single interface for PRs, commits, open or closed, threads on Jira, the MLs, and SO, and a filter for things like my answer to the $1000 question.

Thanks for sharing knowledge.  I have so little, every drop counts, even what doesn't seep in.

**Comment by adriaanm** [comment]
> The Door Is Always Open!

And that's exactly the problem. The more data, the less information.
For all practical purposes, once the PR is closed, it becomes just like the thousands of other closed PRs.

**Comment by paulp** [comment]
We've all taken a look at the content of this PR? I have a lot of trouble warming up to the waiting for waiting's sake. This PR has all the excitement of a spelling correction. If you want a 48-hour minimum or whatever, then so be it. But as long as I have any discretion, I'm not going to leave slam dunk, "spelling correction" level PRs lying around clogging the queue. I don't have it in my DNA.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/Types.scala:2432]: Also, another bonus is that we have nice names for those classes. I was struggling with anonymous classes when working on
https://github.com/gkossakowski/scalac-aspects#type-completion-timings-typecompletiontimingaj where I record class name of completed type.

Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/Types.scala:2432]: Oh, I see that @retronym mentioned that benefit in his description already.

**Comment by gkossakowski** [comment]
@paulp: I agree with you that this PR is trivial to review. My complaint was sent from the device where I didn't even have chance to look at @retronym's patch. It looks good to me as well and it seems like I won't get this $1000 today.

@retronym: do you want to measure shallow size, class instance size or something else? I did various things for different imaginable memory measurements but I agree we should have something methodological, automatic and shared between us. However, I believe that the most useful measures are cumulative ones. Once you clarify what are you after we can come up with something good.

**Pull request merged**